### PR TITLE
rbac: add filter state object option

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Add another option to read principal from the filter state instead of metadata. The use case is passing the value from a network filter to an HTTP filter (RBAC), and the dynamic metadata does not support inheritance. For tunneled requests, the principal needs to be set at the tunnel connection, not the internal connection used for HTTP processing.
Risk Level: low
Testing: unit
Docs Changes: none
Release Notes: none

### CODE REVIEW

**Comment by phlax** [comment]
@yanavlasov would you mind giving this a pass, thanks

**Comment by kyessenov** [comment]
@markdroth Can you please take a look at the API change?

File: source/common/common/matchers.cc [hunk 1]
```diff
@@ -96,6 +96,28 @@ MetadataMatcher::MetadataMatcher(const envoy::type::matcher::v3::MetadataMatcher
   value_matcher_ = ValueMatcher::create(v);
 }
 
+FilterStateMatcher::FilterStateMatcher(const envoy::type::matcher::v3::FilterStateMatcher& matcher)
+    : key_(matcher.key()) {
+  switch (matcher.matcher_case()) {
+  case envoy::type::matcher::v3::FilterStateMatcher::MatcherCase::kStringMatch:
```
Comment (yanavlasov) [inline source/common/common/matchers.cc:102]: Small suggestion that can improve const correctness. You can move this switch statement into a method in anonymous namespace, that returns StringMatcherPtr and then you can use it to initialize value_matcher_ in the constructor initializer list and make it const. I find it easier to reason that configuration objects are immutable after they were created.

**Review by yanavlasov — APPROVED** [review]

File: source/common/common/matchers.cc (hunk 1 above)
Comment (kyessenov) [reply source/common/common/matchers.cc:103]: Good point, immutability also helps compilers.

**Review by yanavlasov — APPROVED** [review]

**Comment by kyessenov** [comment]
Gentle ping @markdroth or any other @envoyproxy/api-shepherds .

**Pull request merged**